File: core/src/main/java/com/function/notification/Notti.java
package com.function.notification;

import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.Context;

import androidx.core.app.NotificationCompat;

import com.function.notification.actions.NotificationAction;
import com.function.notification.config.LightSettings;
import com.function.notification.config.NottiConf;
import com.function.notification.config.VibrationSettings;
import com.function.notification.notifications.CustomNotification;

import java.util.HashMap;
import java.util.Map;

import vn.loitp.utils.util.AppUtils;

public class Notti {

    private Context context;
    private NottiConf nottiConf;
    private NotificationManager notificationManager;
    private Map<String, Integer> ids = new HashMap<>();
    private int currentID = 0;

    public Notti(Context context, NottiConf nottiConf) {
        this.context = context;
        this.nottiConf = nottiConf;
        this.notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
    }

    public void show(CustomNotification customNotification) {
        int notificationId = 1;
        String channelName = AppUtils.getAppName();
        String channelId = "channel-loitp" + channelName;
        int importance = NotificationManager.IMPORTANCE_HIGH;

        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            NotificationChannel mChannel = new NotificationChannel(channelId, channelName, importance);
            notificationManager.createNotificationChannel(mChannel);
        }

        NotificationCompat.Builder builder = new NotificationCompat.Builder(context, channelId);
        setBuilderWithConfig(builder);
        customNotification.setBuilder(builder);
        setActionsForNotification(builder, customNotification);
        setContentIntent(customNotification, builder);
        setDiode(customNotification, builder);
        setVibrations(customNotification, builder);

        notificationManager.notify(getNotificationID(), builder.build());
    }

    private void setVibrations(CustomNotification customNotification, NotificationCompat.Builder builder) {
        VibrationSettings vibrationSettings = nottiConf.getVibrationSettings();
        VibrationSettings customVibrationSettings = customNotification.getVibrationSettings();
        if (vibrationSettings != null && vibrationSettings.isVibrate()) {
            builder.setVibrate(vibrationSettings.getPattern());
        } else if (customVibrationSettings != null && customVibrationSettings.isVibrate()) {
            builder.setVibrate(customVibrationSettings.getPattern());
        }
    }

    private void setDiode(CustomNotification customNotification, NotificationCompat.Builder builder) {
        LightSettings lightSettings = nottiConf.getLightSettings();
        LightSettings customLightSettings = customNotification.getLightSettings();
        if (lightSettings != null) {
            builder.setLights(lightSettings.getArgb(), lightSettings.getOnMs(),
                    lightSettings.getOffMs());
        } else if (customLightSettings != null) {
            builder.setLights(customLightSettings.getArgb(), customLightSettings.getOnMs(),
                    customLightSettings.getOffMs());
        }
    }

    private void setContentIntent(CustomNotification customNotification, NotificationCompat.Builder builder) {
        if (customNotification.getContentAction() != null) {
            builder.setContentIntent(customNotification.getContentAction().getPendingIntent());
        }
    }

    private int provideIcon(Integer icon) {
        if (null == icon) {
            return nottiConf.getDefaultActionImage();
        }
        return icon;
    }

    private void setActionsForNotification(NotificationCompat.Builder builder, CustomNotification customNotification) {
        if (customNotification.getActions() != null) {
            for (NotificationAction notificationAction : customNotification.getActions()) {
                builder.addAction(provideIcon(notificationAction.getImage()),
                        notificationAction.getText(), notificationAction.getPendingIntent());
            }
        }
    }

    public int getNotificationID() {
        if (!nottiConf.isSameID()) {
            return currentID++;
        }
        return 0;
    }

    public int getIDbyString(String string) {
        return ids.get(string);
    }

    public void putID(String string) {
        // TODO: 28.06.16 maybe take random and check if already exists?
        ids.put(string, getNotificationID());
    }

    public void setBuilderWithConfig(NotificationCompat.Builder builder) {
        builder.setSmallIcon(nottiConf.getDefaultActionImage());
    }
}

File: core/src/main/java/com/loitpcore/function/notification/Notti.kt
package com.loitpcore.function.notification

import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.os.Build
import androidx.core.app.NotificationCompat
import com.loitpcore.function.notification.config.NottiConfig
import com.loitpcore.function.notification.notifications.CustomNotification
import com.loitpcore.utils.util.AppUtils

/**
 * Created by Loitp on 04,August,2022
 * Galaxy One company,
 * Vietnam
 * +840766040293
 * freuss47@gmail.com
 */
class Notti(
    private val context: Context,
    private val nottiConfig: NottiConfig
) {

    private val notificationManager: NotificationManager =
        context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
    @Suppress("unused")
    private val ids: MutableMap<String, Int> = HashMap()
    private var currentID = 0

    fun show(customNotification: CustomNotification) {
        val channelName = AppUtils.appName
        val channelId = "channel-$channelName"

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val importance = NotificationManager.IMPORTANCE_HIGH
            val mChannel = NotificationChannel(channelId, channelName, importance)
            notificationManager.createNotificationChannel(mChannel)
        }

        val builder = NotificationCompat.Builder(context, channelId)
        builder.priority = NotificationCompat.PRIORITY_MAX
        setBuilderWithConfig(builder = builder)
        customNotification.setBuilder(builder)
        setActionsForNotification(builder = builder, customNotification = customNotification)
        setContentIntent(customNotification = customNotification, builder = builder)
        setDiode(customNotification = customNotification, builder = builder)
        setVibrations(customNotification = customNotification, builder = builder)
        notificationManager.notify(notificationID, builder.build())
    }

    private fun setVibrations(
        customNotification: CustomNotification,
        builder: NotificationCompat.Builder
    ) {
        val vibrationSettings = nottiConfig.vibrationSettings
        val customVibrationSettings = customNotification.vibrationSettings
        if (vibrationSettings != null && vibrationSettings.isVibrate) {
            builder.setVibrate(vibrationSettings.pattern)
        } else if (customVibrationSettings != null && customVibrationSettings.isVibrate) {
            builder.setVibrate(customVibrationSettings.pattern)
        }
    }

    private fun setDiode(
        customNotification: CustomNotification,
        builder: NotificationCompat.Builder
    ) {
        val lightSettings = nottiConfig.lightSettings
        val customLightSettings = customNotification.lightSettings
        if (lightSettings != null) {
            builder.setLights(lightSettings.argb, lightSettings.onMs, lightSettings.offMs)
        } else if (customLightSettings != null) {
            builder.setLights(
                customLightSettings.argb,
                customLightSettings.onMs,
                customLightSettings.offMs
            )
        }
    }

    private fun setContentIntent(
        customNotification: CustomNotification,
        builder: NotificationCompat.Builder
    ) {
        if (customNotification.contentAction != null) {
            builder.setContentIntent(customNotification.contentAction.getPendingIntent())
        }
    }

    private fun provideIcon(icon: Int): Int {
        return icon
    }

    private fun setActionsForNotification(
        builder: NotificationCompat.Builder,
        customNotification: CustomNotification
    ) {
        if (customNotification.actions != null) {
            for (notificationAction in customNotification.actions) {
                builder.addAction(
                    provideIcon(icon = notificationAction.image),
                    notificationAction.text,
                    notificationAction.getPendingIntent()
                )
            }
        }
    }

    private val notificationID: Int
        get() = if (!nottiConfig.isSameID) {
            currentID++
        } else {
            0
        }

    private fun setBuilderWithConfig(builder: NotificationCompat.Builder) {
        builder.setSmallIcon(nottiConfig.defaultActionImage)
    }
}
